### ISSUE TITLE
visibility of certain content item in different bags

### ISSUE THREAD

**Issue by LatinaAtanasova** [OrchardCMS/OrchardCore#3756]
Hi again,

I have the following senario - bagpart is used as something like "folder" to hold pages with same content type for a certain menu item.

![image](https://user-images.githubusercontent.com/32022885/58884547-53074c80-86e9-11e9-863f-03d491259b8d.png)

For example under "About us" and "Services" there are "pages" of same content type but in order to be client- friendly :) i thought of separating them in two bags - "AboutUsPages" and "ServicesPages". But, last two let's call them folders ;) are different content types. My idea is to use one Razor page to render them by alias. And here comes But... :)
when i try to load them by content type I receive null because they are in a bag. Also they are not visible in ContentPicker Field (one can see only those that are created as new and are not in a bag.

![image](https://user-images.githubusercontent.com/32022885/58885088-3d465700-86ea-11e9-9ea5-b5cc2db534e5.png)

As you can see Documents and About... are not visible (they are included in Bag).
So what would be the solution. I know there is a filter for content Item, but is this the only solution? 
In short:
Is there a way to get all content items of certain content type no matter in what other content types they are part of?

**Comment by deanmarcussen** [comment]
I'm not sure if you're trying to display both these Documents and About pages as separate pages or in one single page. The BagPart would suit if you are wanting to display them all contained in one page, like is done with the Agency theme.

When I have wanted to contain items similar to this, but display them on different pages, i.e. different routes, but under the same section, I have not used the BagPart, but used the ListPart, much like what is done with the Blog theme.

So that makes the Documents page and About page content items, contained in a ListPart by an 'About Us' page. (sometimes I make them listable, sometimes I don't and then they are only accessable by clicking on the parent ListPart in the admin. )

But I don't understand if the 'About Us' section is a page in and of itself, or if it is just a menu title?

If it was just a menu heading then it might suit better to categorize the content in your sections with taxonomies, and if you want to be client friendly in the Admin build an AdminMenu that reflects the front end menu arrangement.

**Comment by LatinaAtanasova** [comment]
Sorry for the explanation :)
"About us" and "Services" are just titles in menu. Under them there will be N-numbers of pages of same type. Their link will be ~/Page/{alias}. So their content will be rendered depending on their alias. On the other hand there will be other content type with content Picker field (to choose from the above N-numbers pages). But these pages are not visible in the content Picker field if they are organized under other content type.
Also it would be nice if a page (content item) can be added as a menu Item Link somehow. I saw it in previous version (Orchard Harvest) and what i found here is that i can add to menuItem Link a Content Picker Field and add the page. And still can't find a way to make the link to take the alias of the chosen page. 

Is there a nice example or tutorial for taxonomy?

**Comment by deanmarcussen** [comment]
Why use BagPart at all?

Would you not create pages of ContentType with an Autoroute so it would be `~/about-metro-expert` and `~/Document` 

If Document is intended to have many different documents in it, then it might have a ListPart so you might have `~/Document` which then shows you a list with summary of all the items in it, and routes might be `~/Document/doc1` and `~/Document/doc2` etc

Have a look at the Blog Theme for how this is done.

Alias is intended to identify documents but I wouldn't imagine seeing it appear in a route anywhere

There's a tutorial on taxonomies here https://www.davidhayden.me/blog/taxonomies-in-orchard-core-cms 

I don't think there is support for a Content Picker Menu Item at this stage, maybe we should have one of those, but for your case use Menu Item Link and use the route of the document you want to link to

**Comment by LatinaAtanasova** [comment]
The only reason for thinking of bag is to use it as "folder" for pages - for clearer view and order of content from user point of view.

**Comment by deanmarcussen** [comment]
> The only reason for thinking of bag is to use it as "folder" for pages - for clearer view and order of content from user point of view.

Lists! definitely the way to get that kind of behaviour. Look at the Blog Theme for examples.

And use Custom Admin menus to make the admin work for you

**Comment by sebastienros** [comment]
I don't disagree with Dean and I have other suggestions that might be more adapted.

First the most obvious solution would have been to use a menu. Having each page as distinct content items, of any type, and only create a hierarchy of menu items in the menu. I understand you tried to make it simpler for the users, but this solution would be the fastest as you don't need to load each content item to render the menu.

Secondly, this is a typical usage of taxonomies, or even simpler, a "Category" text field with predefined values. Then when a user creates a new page, they can associate it with the "topic" they want. A disadvantage is that it is harder to render in the view.

I would opt for the first solution honestly, based on the Menu. Dean's is also simple, each category is a List, and users add pages in these lists, of any allowed content type. You can even reorder them with Dean's recent PR (which hasn't been merged yet, my bad).

Maybe a way to simplify the user experience would be to create feature that would render the existing menu in the editor, such that a user can point if and where this item should be in the menu. Or you just explain to the users that the page will only be in the menu if added into it.

**Comment by LatinaAtanasova** [comment]
Thank you very much for time spent on my dilemmas :) I'll try them and will see what will pass to the user best :)

**Comment by deanmarcussen** [comment]
@sebastienros has summed it the various options much better than I did, and also introduced the possibility of a new feature, which sounds interesting.

Something like a new action button `Add to menu` that could have a drop down if there were multiple menus defined, and auto create a menu item, then render the menu in a modal so the content creator can place it in the correct location in the menu.

Is that what you were suggesting Sebastien?

Another question raised by @LatinaAtanasova was if there was support for a 'Content Menu Link` to maintain a link directly to a content item rather than a route.

I know in O1 that could cause N+1 issues when the number of items in the menu grew, so am questioning whether it's a desired feature to add?

We could probably cache it well, and clear the cache on Content Published

**Comment by LatinaAtanasova** [comment]
@deanmarcussen I just want to say big thank you for the List :) It is definitely perfect for my idea and i did one more thing - To linkmenuitem added a Content Picker Field. So in back end iterate over all menu items and in case there are sub items with chosen "page" i take their link and title for the navigation. I don't know if this is the best approach but in this case the user can create item links and links to content items just with the limitation of content picker. 
I forgot to say that i use the decoupled approach. Thank you both @deanmarcussen @sebastienros again :)

**Comment by sebastienros** [comment]
Yes, the Content Menu Link feature worries me because it was a big perf issue in O1. the menu is cached, so maybe not that of an issue. Or the feature could store the links only, intercept changes to content items and update them if necessary.

**Comment by deanmarcussen** [comment]
> Yes, the Content Menu Link feature worries me because it was a big perf issue in O1. 

Yes, it's a challenge a Content Menu Link feature, no matter how well cached it is, it still always needs to build itself before being cacheable.

@LatinaAtanasova For infos, the problem with using a Content Picker Field is that it must retrieve each ContentItem from the database, one at a time, every time it renders the menu (until it caches it), so it can be a performance bottleneck, which you may not notice until the site and menu grows.

For me, when I've wanted to build dynamic navigation automatically, with a large quantity of deeply nested items in it, I've bypassed the Menu module entirely, and implemented an `INavigationProvider` (see the way the admin menu is built) which I can populate either with a single `ISession.Query` to an Index, or with a taxonomy query like Sebastien's suggestion above. Then the query can be cached, until content is updated, and the navigation can then be rendered with a `Navigation` shape. 
So far this technique has proved quite successful. I think I wrote it before taxonomies was available, but if I did it again, I'd use taxonomies.

Regardless the `Add to menu` feature sounds cool. (maybe for the backlog?)
